src/PostgREST/Error.hs

```diff
@@ -343,3 +410,81 @@ invalidTokenHeader m =
 
 singularityError :: (Integral a) => a -> Error
 singularityError = SingularityError . toInteger
+
+-- Error codes are grouped by common modules or characteristics
+data ErrorCode
+  -- PostgreSQL connection errors
+  = ConnectionErrorCode00
+  | ConnectionErrorCode01
+  | ConnectionErrorCode02
+  -- API Request errors
+  | ApiRequestErrorCode00
+  | ApiRequestErrorCode01
+  | ApiRequestErrorCode02
+  | ApiRequestErrorCode03
+  | ApiRequestErrorCode04
+  | ApiRequestErrorCode05
+  | ApiRequestErrorCode06
+  | ApiRequestErrorCode07
+  -- Schema Cache errors
+  | SchemaCacheErrorCode00
+  | SchemaCacheErrorCode01
+  | SchemaCacheErrorCode02
+  | SchemaCacheErrorCode03
+  -- JWT authentication errors
+  | JWTErrorCode00
+  | JWTErrorCode01
+  -- Hasql library errors
+  | HasqlErrorCode00
+  | HasqlErrorCode01
+  | HasqlErrorCode02
+  | HasqlErrorCode03
+  | HasqlErrorCode04
+  -- Uncategorized errors that are not related to a single module
+  | GeneralErrorCode00
+  | GeneralErrorCode01
+  | GeneralErrorCode02
+  | GeneralErrorCode03
+  | GeneralErrorCode04
+  | GeneralErrorCode05
+
+instance JSON.ToJSON ErrorCode where
+  toJSON e = JSON.toJSON (buildErrorCode e)
+
+-- New group of errors will be added at the end of all the groups and will have the next prefix in the sequence
+-- New errors are added at the end of the group they belong to and will have the next code in the sequence
+buildErrorCode :: ErrorCode -> Text
+buildErrorCode code = "PGRST" <> case code of
+  ConnectionErrorCode00  -> "000"
+  ConnectionErrorCode01  -> "001"
+  ConnectionErrorCode02  -> "002"
+
+  ApiRequestErrorCode00  -> "100"
+  ApiRequestErrorCode01  -> "101"
+  ApiRequestErrorCode02  -> "102"
+  ApiRequestErrorCode03  -> "103"
+  ApiRequestErrorCode04  -> "104"
+  ApiRequestErrorCode05  -> "105"
+  ApiRequestErrorCode06  -> "106"
+  ApiRequestErrorCode07  -> "107"
+
+  SchemaCacheErrorCode00 -> "200"
+  SchemaCacheErrorCode01 -> "201"
+  SchemaCacheErrorCode02 -> "202"
+  SchemaCacheErrorCode03 -> "203"
+
+  JWTErrorCode00         -> "300"
+  JWTErrorCode01         -> "301"
+
+  HasqlErrorCode00       -> "400"
+  HasqlErrorCode01       -> "401"
+  HasqlErrorCode02       -> "402"
+  HasqlErrorCode03       -> "403"
+  HasqlErrorCode04       -> "404"
```

Comment:
Codecov is giving a warning for these errors codes and their error instances, but I don't think they can be tested because they are caused by unexpected bugs. We might have to tank the score loss here.